lms/resources/_js_config.py

```diff
@@ -16,16 +16,23 @@ class JSConfig:  # pylint:disable=too-many-instance-attributes
     def __init__(self, context, request):
         self._context = context
         self._request = request
-        self._lti_user = request.lti_user
-        self._h_user = request.lti_user.h_user
-        self._consumer_key = (
-            request.lti_user.oauth_consumer_key if request.lti_user else None
-        )
         self._authority = request.registry.settings["h_authority"]
-
         self._grading_info_service = request.find_service(name="grading_info")
-        self._ai_getter = request.find_service(name="ai_getter")
         self._h_api = request.find_service(name="h_api")
+        self._lti_user = request.lti_user
+
+    @property
+    def _h_user(self):
+        return self._lti_user.h_user
+
+    @property
+    def _consumer_key(self):
+        return self._lti_user.oauth_consumer_key
+
+    @property
+    def _ai_getter(self):
+        # nb. `ai_getter` service is only available in LTI launches.
```

Comment:
```suggestion
```

This comment belongs in a docstring: it's information that the person calling the method wants to know. "Restrictions on when it can be called" are one of the things that belong in a multiline method docstring according to PEP 257.

Also, the property is available during all authenticated requests not just LTI launches.

The same restriction applies to the `_h_user()` and `_consumer_key()` properties above.

I think simply deleting the comment would also be acceptable since they're private methods (but any public methods that call these should note the restriction in their docstrings).

Reply:
> Also, the property is available during all authenticated requests not just LTI launches.

I'm wary of writing that because there may be requests in future where there the caller is "authenticated" in some way but `request.lti_user` is not set. For now I've just removed this comment since, as you noted, it is not entirely correct.